src/rules/restrict-dependency-ranges.ts

```diff
@@ -0,0 +1,293 @@
+import type { AST as JsonAST } from "jsonc-eslint-parser";
+
+import * as ESTree from "estree";
+import semver from "semver";
+
+import { createRule } from "../createRule.js";
+import { isJSONStringLiteral } from "../utils/predicates.js";
+
+const DEPENDENCY_TYPES = [
+	"dependencies",
+	"devDependencies",
+	"optionalDependencies",
+	"peerDependencies",
+];
+
+interface Option {
+	forDependencyTypes?: string[];
+	forPackages?: string[];
+	forVersions?: string;
+	rangeType: RangeType | RangeType[];
+}
+
+type Options = [Option | Option[] | undefined];
+
+const RANGE_TYPES = ["caret", "pin", "tilde"];
+type RangeType = "caret" | "pin" | "tilde";
```

Comment:
[Style] Nit, so the code doesn't have to repeat itself:

```suggestion
const RANGE_TYPES = ["caret", "pin", "tilde"] as const;
type RangeType = (typeof RANGE_TYPES)[number];
```

Reply:
Updated